clang/lib/AST/ExprConstant.cpp

```diff
@@ -10895,6 +10899,132 @@ bool VectorExprEvaluator::VisitUnaryOperator(const UnaryOperator *E) {
+  if (!Evaluate(Result, Info, E))
+    return false;
+
+  if (Result.isLValue()) {
+    // Source of the data is an lvalue; Manually handle the lvalue as if
+    // it was an rvalue to get the current APValue.
+    LValue LValueFound;
+    LValueFound.setFrom(Info.Ctx, Result);
+    if (!handleLValueToRValueConversion(Info, E, Type, LValueFound, Result))
+      return false;
+  }
+
+  return Result.isVector();
+}
+
+static bool handleVectorConversion(EvalInfo &Info, const FPOptions FPO,
+                                   const Expr *E, QualType SourceTy,
+                                   QualType DestTy, APValue const &Original,
+                                   APValue &Result) {
+  if (SourceTy->isIntegerType()) {
+    if (DestTy->isRealFloatingType()) {
+      Result = APValue(APFloat(0.0));
+      return HandleIntToFloatCast(Info, E, FPO, SourceTy, Original.getInt(),
+                                  DestTy, Result.getFloat());
+    }
+    if (DestTy->isIntegerType()) {
+      Result = APValue(
+          HandleIntToIntCast(Info, E, DestTy, SourceTy, Original.getInt()));
+      return true;
+    }
+  } else if (SourceTy->isRealFloatingType()) {
+    if (DestTy->isRealFloatingType()) {
+      Result = Original;
+      return HandleFloatToFloatCast(Info, E, SourceTy, DestTy,
+                                    Result.getFloat());
+    }
+    if (DestTy->isIntegerType()) {
+      Result = APValue(APSInt());
+      return HandleFloatToIntCast(Info, E, SourceTy, Original.getFloat(),
+                                  DestTy, Result.getInt());
+    }
+  }
+  return false;
+}
+
+bool VectorExprEvaluator::VisitConvertVectorExpr(const ConvertVectorExpr *E) {
+  APValue Source;
+  QualType SourceVecType = E->getSrcExpr()->getType();
+  if (!EvaluateVectorOrLValue(Source, Info, E->getSrcExpr(), SourceVecType))
+    return false;
+
+  QualType DestTy = E->getType()->castAs<VectorType>()->getElementType();
+  QualType SourceTy = SourceVecType->castAs<VectorType>()->getElementType();
+
+  const FPOptions FPO = E->getFPFeaturesInEffect(Info.Ctx.getLangOpts());
+
+  auto SourceLen = Source.getVectorLength();
+  SmallVector<APValue, 4> ResultElements;
+  ResultElements.reserve(SourceLen);
+  for (unsigned EltNum = 0; EltNum < SourceLen; ++EltNum) {
+    APValue Elt;
+    if (!handleVectorConversion(Info, FPO, E, SourceTy, DestTy,
+                                Source.getVectorElt(EltNum), Elt))
+      return false;
+    ResultElements.push_back(std::move(Elt));
+  }
+
+  return Success(APValue(ResultElements.data(), ResultElements.size()), E);
+}
+
+static bool handleVectorShuffle(EvalInfo &Info, const ShuffleVectorExpr *E,
+                                QualType ElemType, APValue const &VecVal1,
+                                APValue const &VecVal2, unsigned EltNum,
+                                APValue &Result) {
+  unsigned const TotalElementsInAVector = VecVal1.getVectorLength();
```

Comment:
Aren't both vectors the same size? This code assumes that based on the spec. If They are not the same size, I need to revisit this, but otherwise I think the name is just fine. Maybe I'll rename it to `TotalElementsInInputVector` so it is a bit more descriptive

Reply:
I didn't think the vectors had to be the same length: both clang and gcc document the extension to require that the vectors have "a compatible element type" ([gcc](https://gcc.gnu.org/onlinedocs/gcc/Vector-Extensions.html)) or "are vectors that have the same element type" ([clang](https://clang.llvm.org/docs/LanguageExtensions.html#builtin-shufflevector)). 

However, while gcc does accept a shuffle of e.g. an 8-element and 4-element vector (for a 12-element "input"), clang seems to reject it: https://godbolt.org/z/3ooszMKbc

So maybe the thing to do here is add an `assert(VecVal1.getVectorLength() == VecVal2.getVectorLength())`? And/or, change the documentation to make it clear it's "same element type _and length_"?

An alternative to consider would be to plumb both the `constexpr` and codegen pieces to accept a wider range of vectors for arguments 1 and 2, but I'm assuming the current lowering is just using LLVM's `shufflevector`, which itself only supports homogeneous arguments. To me, that smells like something best addressed as its own unit of work and far outside the scope of this PR, but I wanted to include it for completeness.

Reply:
I see! It seems like Gcc has extra characteristics then. given that at some point it could happen that such a thing needs to be supported, let's actually rewrite the code to support both sizes. Beware, I won't be able to test it, as Sema will prevent it from reaching constexpr.